File: scripts/raw/lib/pdf-extract.py
# Helper d'extraction texte page-a-page pour scripts/raw/anchor-fill.mjs (#522).
# Lit le PDF UNE SEULE FOIS (pypdf), extrait le texte des pages demandees (index pypdf 0-based,
# = id "page-K-0" pose par Marker), ecrit un JSON {K: texte} sur stdout.
# Usage : python pdf-extract.py <pdf-path> <K1,K2,...> <out-json-path>
# Ecrit en UTF-8 dans un FICHIER (jamais stdout : la console Windows par defaut est cp1252 et
# rejette les glyphes hors-table que pypdf peut produire sur du texte mal encode dans le PDF source).
import sys
import json
from pypdf import PdfReader

def main():
    pdf_path = sys.argv[1]
    indices = [int(x) for x in sys.argv[2].split(',') if x != '']
    out_path = sys.argv[3]
    reader = PdfReader(pdf_path)
    out = {}
    for k in indices:
        if k < 0 or k >= len(reader.pages):
            out[str(k)] = None
            continue
        text = reader.pages[k].extract_text() or ""
        out[str(k)] = text
    with open(out_path, 'w', encoding='utf-8') as f:
        json.dump(out, f, ensure_ascii=False)

if __name__ == '__main__':
    main()

File: scripts/raw/lib/pdf-extract.mjs
// Extraction TEXTE d'un PDF par pypdf (`pdf-extract.py`), lecteur des pages perdues :
// `empty-folios-stock.mjs`, `marker-pages.mjs`.
import { mkdtempSync, readFileSync, rmSync } from 'node:fs'
import { dirname, join } from 'node:path'
import { tmpdir } from 'node:os'
import { fileURLToPath } from 'node:url'
import { execFileSync } from 'node:child_process'

export const PDF_EXTRACT_SCRIPT = join(dirname(fileURLToPath(import.meta.url)), 'pdf-extract.py')

/** Texte des pages `indices` (0-based) de `pdfPath` : `Map(index -> texte)`, un seul process python. */
export function extractPages(pdfPath, indices) {
  if (!indices.length) return new Map()
  const dir = mkdtempSync(join(tmpdir(), 'pdf-extract-'))
  const outPath = join(dir, 'pages.json')
  try {
    execFileSync('python', [PDF_EXTRACT_SCRIPT, pdfPath, indices.join(','), outPath], { maxBuffer: 128 * 1024 * 1024 })
    const raw = JSON.parse(readFileSync(outPath, 'utf8'))
    const map = new Map()
    for (const [k, v] of Object.entries(raw)) map.set(Number(k), v)
    return map
  } finally {
    rmSync(dir, { recursive: true, force: true })
  }
}
